### PR TITLE
bpo-35017: Lib/socketserver, do not accept any request after shutdown

### PR DESCRIPTION
**Pull request by beledouxdenis**
Prior to this revision,
After the shutdown of a `BaseServer`,
the server accepted a last single request
if it was sent between the server socket polling
and the polling timeout.

This can be problematic for instance
for a server restart
for which you do not want to interrupt the service,
by not closing the listening socket during the restart.
One request failed because of this behavior.

Note that only one request failed,
following requests were not accepted, as expected.

https://bugs.python.org/issue35017

We solved this issue in [odoo/odoo](https://github.com/odoo/odoo) by overriding _handle_request_noblock, in the below revision:
https://github.com/odoo/odoo/commit/dcb56412d667231c5427b0cd76cfc1c6c4b1dd3c#diff-7c4c00ad10cee1a79d1ed0d8b96f0bfaR140

We would like to avoid having to do so, as this is not a method which is supposed to be overridden.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Misc/NEWS.d/next/Library/2018-10-19-09-32-07.bpo-35017.vAanVd.rst
Comment (tirkarthi) [inline Misc/NEWS.d/next/Library/2018-10-19-09-32-07.bpo-35017.vAanVd.rst]: As a personal suggestion I think this can be reworded to represent the behavior the PR proposes instead of the current behavior like below. I would wait for reviewer feedback though. 

```markdown
:class:`BaseServer` rejects the request sent between polling and polling 
timeout after shutdown.
```

Other than that it looks good to me. Thanks for the PR :)

**Comment by tirkarthi** [comment]
I think Victor might help here in reviewing this PR. Friendly ping @vstinner

File: Lib/socketserver.py
Comment (vstinner) [inline Lib/socketserver.py:232]: Why not "if not self.__shutdown_request: break"? Is it useful to call self.service_actions() even if we asked to stop the server? If you add this break, please add a short comment to explain how it can happen and add a "bpo-35017: " prefix to your comment to add the context.

Comment (beledouxdenis) [reply Lib/socketserver.py:232]: You are right. I will make the change right away.

**Comment by beledouxdenis** [comment]
Thank you very much for your review @vstinner 
I have updated the PR with your remarks.

This targets a bug in Python 3.6 and 3.7. Is there something additional to do for this revision to land in these Python releases or will it be back-ported later, potentially through another pull request ?

**Comment by tirkarthi** [comment]
@beledouxdenis Once the PR is merged and the reviewer adds "needs backport to 3.7" and "needs backport to 3.6" labels to the PR @miss-islington will cherry pick the commit and create relevant PRs to the branches. If there are any conflicts then you might need to create a PR manually fixing the conflicts during cherry picks. Given the change I think there will be no conflicts.

Relevant dev guide section : https://devguide.python.org/committing/?highlight=backport#backporting-changes-to-an-older-version

**Comment by beledouxdenis** [comment]
Thank you for the clarification !

**Comment by tirkarthi** [comment]
You're welcome :) I could see only one commit for the PR. In case you have force pushed changes it will make review of the changes between review comments little harder to notice unlike having separate commits. Core devs squash and merge all the commits in a PR as a single commit so for future contributions please push the changes as separate commits for reviews.

Thanks

File: Lib/test/test_socketserver.py
Comment (vstinner) [inline Lib/test/test_socketserver.py]: It seems like your test rely on time for synchronization. I feel that it will fail if the sytem is busy or the threads are ordered differently depending on the OS :-(

I would either prefer to remove the test or write a test with better synchronization between the 3 threads.

cc @pablogsal

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/test/test_socketserver.py
Comment (beledouxdenis) [reply Lib/test/test_socketserver.py]: The unit test was mainly to demonstrate there was an issue and how it could be triggered.

I hardly see how I could make the unit test not time based.

For the part
```python
sock.settimeout(0.1)
...
try:
    sock.recv(1024)
except socket.timeout:
    rejected = True
```
I could indeed maybe find another heuristic to determine the request was not accepted, but then remains the ```poll_interval``` timeout of the method
```python
def serve_forever(self, poll_interval=0.5):
```

If I increase this, this will take longer for the server to shutdown, and therefore the test will take time for an unnecessary reason.

I have no problem the unit test not to be merged. As explained, this was mainly to demonstrate the issue. Maybe it can be used as a reference in this pull request in case someone would like to know why this two lines patch was added.

I am opened to your suggestions if you have an idea to make this not time-based. Otherwise let me know and I will add a commit to remove the unit test.

Comment (pablogsal) [reply Lib/test/test_socketserver.py]: Indeed, I was able to make this test fail several times in our slowest buildbots :(

There is even the possibility that `shutdown` is called before `server_forver` starts the loop.

Also, shouldn't this be decorated with `@reap_threads`?

Comment (vstinner) [reply Lib/test/test_socketserver.py]: Please just remove the test. The fix is trivial, but I don't want to have to debug race condition in this new functional test.

File: Misc/NEWS.d/next/Library/2018-10-19-09-32-07.bpo-35017.vAanVd.rst
Comment (vstinner) [reply Misc/NEWS.d/next/Library/2018-10-19-09-32-07.bpo-35017.vAanVd.rst]: Please mention the module name. Something like:

```
:meth:`socketserver.BaseServer.serve_forever` now exits immediately
if it's :meth:`~socketserver.BaseServer.shutdown` method is called
while it is polling for new events.
```

File: Lib/test/test_socketserver.py
Comment (beledouxdenis) [reply Lib/test/test_socketserver.py]: I fully agree. I just removed the unit test case.

File: Misc/NEWS.d/next/Library/2018-10-19-09-32-07.bpo-35017.vAanVd.rst
Comment (beledouxdenis) [reply Misc/NEWS.d/next/Library/2018-10-19-09-32-07.bpo-35017.vAanVd.rst]: I removed the former news entry and added a new one with your description. I couldn't have described this better.

File: Lib/socketserver.py
Comment (vstinner) [inline Lib/socketserver.py]: I'm confused by this comment. I suggest something like:

"bpo-35017: shutdown() has been called from a different thread while select() was polling. Don't handle the request, exit immediatly."

... or maybe remove the comment, I don't know :-)

Comment (odony) [reply Lib/socketserver.py]: If I may suggest a shorter version with the same meaning:
```python
# bpo-35017: shutdown() called during select(), exit immediately
``` 
Though I think the code is quite self-explanatory, and I would omit the comment ;-)

Comment (beledouxdenis) [reply Lib/socketserver.py]: Let me know If I should use @odony  reword, or if you prefer removing the comment completely.

Comment (vstinner) [reply Lib/socketserver.py]: "# bpo-35017: shutdown() called during select(), exit immediately" is better than my long boring comment :)

Comment (tirkarthi) [reply Lib/socketserver.py]: @vstinner Off topic but just as a pro-tip now you can suggest changes in GitHub UI that can be accepted and committed from UI when the contributor clicks "Accept suggestion" that might help during rewording :) Ref : https://twitter.com/github/status/1054794897448939520 

Edit : Doesn't work with multiple lines it seems. Sorry for the noise.

```suggestion
                    # bpo-35017: shutdown() called during select(), exit immediately
```

**Review by vstinner — APPROVED** [review]

**Comment by vstinner** [comment]
@beledouxdenis: Hum, the bot failed to create a backport for Python 2.7. Can you please try to do the backport manually? Use "cherry_picker 10cb3760e8631a27f5db1e51b05494e29306c671 2.7" or "git cherry-pick -x 10cb3760e8631a27f5db1e51b05494e29306c671".

**Comment by beledouxdenis** [comment]
I didn't expect the backport to 2.7, I am not even sure the bug is there. I initially targeted 3.6 and 3.7.

Will check anyway if this can be reproduced in 2.7 :).

**Comment by vstinner** [comment]
> I didn't expect the backport to 2.7, I am not even sure the bug is there.

The code is almost the same, so I'm sure that it has the bug:

```
while not self.__shutdown_request:
    # XXX: Consider using another file descriptor or
    # connecting to the socket to wake this up instead of
    # polling. Polling reduces our responsiveness to a
    # shutdown request and wastes cpu at all other times.
    r, w, e = _eintr_retry(select.select, [self], [], [],
                           poll_interval)
    if self in r:
        self._handle_request_noblock()
```

The fix should be the same :-)

**Comment by beledouxdenis** [comment]
I came to the same conclusion. I can re-use my unit test after all :-).

**Pull request merged**
